### PR TITLE
General preparation for fuller sibra support.

### PR DESCRIPTION
**Pull request by kormat**
Router changes:
- Read signing key on startup
- Pass from_local_ad flag to extension handling
- Allow deliver() to work if there's no scion path
- Extract egress forwarding into seperate function, so it can be called
  by _process_flags().

Elem:
- Update get_first_hop to allow extensions to provide the interface.
- Check that src/dst AD are different when there's no path (and an
  extension hasn't provided the interface).

Misc:
- Move extension types into lib.types
- Add support for reversing extensions, and getting interface IDs from
  extensions.
- Adding utility function to find an extension of a specific class/type.
- EmptyPath.get_hof() now returns None
- Enforce extension headers to be a multiple of LINE_LEN
- Allow setting of timeouts on sockets
  <a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/netsec-ethz/scion/pull/590%23issuecomment-172473759%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/590%23discussion_r50235710%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/590%23discussion_r50235989%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/590%23issuecomment-173159244%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/590%23discussion_r51238116%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/590%23discussion_r51238144%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/590%23discussion_r51392630%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/590%23discussion_r51393022%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/590%23discussion_r51394366%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/590%23discussion_r51394692%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/590%23issuecomment-172473759%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%40pszalach%20%22%2C%20%22created_at%22%3A%20%222016-01-18T09%3A28%3A08Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22lgtm%20%28but%20it%20makes%20sense%20to%20me%20to%20review%20bigger%20parts%29%22%2C%20%22created_at%22%3A%20%222016-01-20T10%3A16%3A26Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20bb1355104800c3adc42abf2221112b8f897562e7%20infrastructure/router/main.py%2027%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/590%23discussion_r50235710%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22what%20does%20router%20sign%3F%22%2C%20%22created_at%22%3A%20%222016-01-20T10%3A10%3A31Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22Sibra%20Opaque%20Fields.%22%2C%20%22created_at%22%3A%20%222016-01-29T09%3A00%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22it%20is%20created%20with%20symmetric%20key%2C%20so%20I%20suggest%20to%20just%20derive%20one%3A%5Cr%5Cn%60self.sibra_key%20%3D%20PBKDF2%28self.config.master_ad_key%2C%20b%5C%22Derive%20SIBRA%20Key%5C%22%29%60%22%2C%20%22created_at%22%3A%20%222016-02-01T09%3A00%3A36Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22Fixed.%22%2C%20%22created_at%22%3A%20%222016-02-01T09%3A20%3A57Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20infrastructure/router/main.py%3AL132-140%22%7D%2C%20%22Pull%20bb1355104800c3adc42abf2221112b8f897562e7%20infrastructure/router/main.py%2074%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/590%23discussion_r50235989%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22router%20should%20not%20accept%20packets%20with%20empty%20path.%20At%20some%20point%20we%20need%20to%20have%20%60_is_spkt_correct%28%29%60%20that%20would%20do%20sanity%20checks.%22%2C%20%22created_at%22%3A%20%222016-01-20T10%3A12%3A53Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22This%20is%20not%20true.%20The%20majority%20of%20sibra%20packets%20will%20have%20empty%20paths.%22%2C%20%22created_at%22%3A%20%222016-01-29T09%3A01%3A02Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20see%2C%20but%20then%20maybe%20router%20should%20check%20that%20%28i.e.%2C%20if%20path%20is%20empty%20then%20sibra%20ext%20has%20to%20exist%29%2C%20...%20at%20least%20please%20comment%20that.%22%2C%20%22created_at%22%3A%20%222016-02-01T09%3A05%3A18Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22Added%20a%20comment.%22%2C%20%22created_at%22%3A%20%222016-02-01T09%3A25%3A02Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20infrastructure/router/main.py%3AL425-433%22%7D%7D%7D'></a>
  <a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/590#issuecomment-172473759'>General Comment</a></b>
- <a href='https://github.com/kormat'></a> @pszalach
- <a href='https://github.com/pszalach'></a> lgtm (but it makes sense to me to review bigger parts)
- [ ] <a href='#crh-comment-Pull bb1355104800c3adc42abf2221112b8f897562e7 infrastructure/router/main.py 27'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/590#discussion_r50235710'>File: infrastructure/router/main.py:L132-140</a></b>
- <a href='https://github.com/pszalach'></a> what does router sign?
- <a href='https://github.com/kormat'></a> Sibra Opaque Fields.
- <a href='https://github.com/pszalach'></a> it is created with symmetric key, so I suggest to just derive one:
  `self.sibra_key = PBKDF2(self.config.master_ad_key, b"Derive SIBRA Key")`
- <a href='https://github.com/kormat'></a> Fixed.
- [ ] <a href='#crh-comment-Pull bb1355104800c3adc42abf2221112b8f897562e7 infrastructure/router/main.py 74'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/590#discussion_r50235989'>File: infrastructure/router/main.py:L425-433</a></b>
- <a href='https://github.com/pszalach'></a> router should not accept packets with empty path. At some point we need to have `_is_spkt_correct()` that would do sanity checks.
- <a href='https://github.com/kormat'></a> This is not true. The majority of sibra packets will have empty paths.
- <a href='https://github.com/pszalach'></a> I see, but then maybe router should check that (i.e., if path is empty then sibra ext has to exist), ... at least please comment that.
- <a href='https://github.com/kormat'></a> Added a comment.

<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/590?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/590?approve=1'></a>&nbsp;<a href='https://github.com/netsec-ethz/scion/pull/590'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: infrastructure/router/main.py [hunk 1]
```diff
@@ -126,6 +132,8 @@ def __init__(self, server_id, conf_dir, is_sim=False):
         self.of_gen_key = PBKDF2(self.config.master_ad_key, b"Derive OF Key")
         self.if_states = defaultdict(InterfaceState)
         self.revocations = ExpiringDict(1000, self.FWD_REVOCATION_TIMEOUT)
+        sig_key_file = get_sig_key_file_path(self.conf_dir)
+        self.signing_key = base64.b64decode(read_file(sig_key_file))
```
Comment (pszal) [inline infrastructure/router/main.py]: what does router sign?

File: infrastructure/router/main.py [hunk 2]
```diff
@@ -416,8 +425,8 @@ def deliver(self, spkt, ptype):
         """
         path = spkt.path
         curr_hof = path.get_hof()
-        if (not path.is_last_path_hof() or
-                (curr_hof.ingress_if and curr_hof.egress_if)):
+        if len(path) and (not path.is_last_path_hof() or
```
Comment (pszal) [inline infrastructure/router/main.py]: router should not accept packets with empty path. At some point we need to have `_is_spkt_correct()` that would do sanity checks.

File: infrastructure/router/main.py (hunk 1 above)
Comment (kormat) [inline infrastructure/router/main.py]: Sibra Opaque Fields.

File: infrastructure/router/main.py (hunk 2 above)
Comment (kormat) [inline infrastructure/router/main.py]: This is not true. The majority of sibra packets will have empty paths.

File: infrastructure/router/main.py (hunk 1 above)
Comment (pszal) [inline infrastructure/router/main.py]: it is created with symmetric key, so I suggest to just derive one:
`self.sibra_key = PBKDF2(self.config.master_ad_key, b"Derive SIBRA Key")`

File: infrastructure/router/main.py (hunk 2 above)
Comment (pszal) [inline infrastructure/router/main.py]: I see, but then maybe router should check that (i.e., if path is empty then sibra ext has to exist), ... at least please comment that.

File: infrastructure/router/main.py (hunk 1 above)
Comment (kormat) [inline infrastructure/router/main.py]: Fixed.

File: infrastructure/router/main.py (hunk 2 above)
Comment (kormat) [inline infrastructure/router/main.py]: Added a comment.

**Pull request merged**